### PR TITLE
Postgres: add simple LargeObject support

### PR DESCRIPTION
**Pull request by stevenschlansker**
https://github.com/jdbi/jdbi/issues/1706

### CODE REVIEW

File: postgres/src/main/java/org/jdbi/v3/postgres/BlobInputStreamArgumentFactory.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.sql.Types;
+
+import org.jdbi.v3.core.argument.AbstractArgumentFactory;
+import org.jdbi.v3.core.argument.Argument;
+import org.jdbi.v3.core.config.ConfigRegistry;
+import org.jdbi.v3.core.statement.StatementContext;
+
+class BlobInputStreamArgumentFactory extends AbstractArgumentFactory<InputStream> {
```
Comment (leaumar) [inline postgres/src/main/java/org/jdbi/v3/postgres/BlobInputStreamArgumentFactory.java:13]: Are there no other potential/reasonable use cases for an InputStream in queries, like are these postgres lobs equivalent to general lobs for all dbs? If there are uses that don't mix with this, this may need a qualifier

File: postgres/src/main/java/org/jdbi/v3/postgres/ClobReaderColumnMapperFactory.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import org.jdbi.v3.postgres.BlobInputStreamColumnMapperFactory.LobColumnMapper;
+
+class ClobReaderColumnMapperFactory implements ColumnMapperFactory {
+    @Override
+    public Optional<ColumnMapper<?>> build(Type type, ConfigRegistry config) {
+        if (Reader.class != type) {
+            return Optional.empty();
+        }
+        LobColumnMapper inner = new LobColumnMapper();
+        return Optional.of((rs, col, ctx) -> new InputStreamReader(inner.map(rs, col, ctx), StandardCharsets.UTF_8));
```
Comment (leaumar) [inline postgres/src/main/java/org/jdbi/v3/postgres/ClobReaderColumnMapperFactory.java:21]: I'm assuming this is similarly hardcoded in postgres then? ctrl-f "utf" on the doc page shows nothing.

File: postgres/src/main/java/org/jdbi/v3/postgres/PgLobImpl.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+import java.io.InputStream;
+import java.sql.Connection;
+import java.sql.SQLException;
+
+import org.postgresql.PGConnection;
+import org.postgresql.largeobject.LargeObject;
+import org.postgresql.largeobject.LargeObjectManager;
+
+class PgLobImpl implements PgLob {
+    private static final int BUF_SIZE = 1024 * 64;
```
Comment (leaumar) [inline postgres/src/main/java/org/jdbi/v3/postgres/PgLobImpl.java:13]: Any particular reason for this exact number? Wouldn't this make a good Config value?

File: postgres/src/main/java/org/jdbi/v3/postgres/PostgresTypes.java [hunk 4]
```diff
@@ -56,6 +59,20 @@ public PostgresTypes registerCustomType(Class<? extends PGobject> clazz, String
         return this;
     }
 
+    PostgresTypes setLobApi(PgLob lob) {
```
Comment (leaumar) [inline postgres/src/main/java/org/jdbi/v3/postgres/PostgresTypes.java:62]: Is the name `PgLob` appropriate? It reads like an instance of a lob, not as the handling implementation for lobs.

File: postgres/src/test/java/org/jdbi/v3/postgres/TestLobStream.java [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.postgresql.PGConnection;
+import org.postgresql.largeobject.LargeObjectManager;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestLobStream {
+    private static final int BIGLY = 1024 * 1024 * 64;
```
Comment (leaumar) [inline postgres/src/test/java/org/jdbi/v3/postgres/TestLobStream.java:35]: I'm missing an expression of what this is for and why it's the value that it is, e.g. `SAMPLE_SIZE_GREATLY_EXCEEDING_LIMIT` or a short comment?

File: postgres/src/test/java/org/jdbi/v3/postgres/TestLobStream.java [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+            assertSameChars(lob.findClob(2), expectedData.get());
+        });
+    }
+
+    private void assertSameBytes(InputStream a, InputStream b) throws IOException {
+        int pos = 0;
+        int read;
+        while ((read = a.read()) != -1) {
+            assertThat(b.read())
+                    .describedAs("byte at position %s", pos)
```
Comment (leaumar) [inline postgres/src/test/java/org/jdbi/v3/postgres/TestLobStream.java:77]: is it a byte though? An int is 4 bytes

File: postgres/src/test/java/org/jdbi/v3/postgres/TestLobStream.java [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+            lob.insert(2, expectedData.get());
+            assertSameChars(lob.findClob(2), expectedData.get());
+        });
+    }
+
+    private void assertSameBytes(InputStream a, InputStream b) throws IOException {
+        int pos = 0;
+        int read;
+        while ((read = a.read()) != -1) {
+            assertThat(b.read())
```
Comment (leaumar) [inline postgres/src/test/java/org/jdbi/v3/postgres/TestLobStream.java:76]: if I understand right, this is gonna be called for 1024*1024*64 ints, so isn't this test... really slow, asserting one int at a time?

File: sqlobject/src/main/java/org/jdbi/v3/sqlobject/statement/internal/ResultReturner.java [hunk 8]
```diff
@@ -252,7 +275,11 @@ protected Object reducedResult(Stream<?> stream, StatementContext ctx) {
         }
     }
 
-    static class CollectedResultReturner<T> extends ResultReturner {
+    public static ResultReturner forCollecetedType(QualifiedType<?> returnType) {
```
Comment (leaumar) [inline sqlobject/src/main/java/org/jdbi/v3/sqlobject/statement/internal/ResultReturner.java:278]: "colleceted"

File: sqlobject/src/main/java/org/jdbi/v3/sqlobject/statement/internal/ResultReturner.java [hunk 9]
```diff
@@ -109,7 +109,11 @@ private static Object checkResult(Object result, QualifiedType<?> type) {
         return result;
     }
 
-    static class VoidReturner extends ResultReturner {
+    public static ResultReturner forVoid() {
```
Comment (leaumar) [inline sqlobject/src/main/java/org/jdbi/v3/sqlobject/statement/internal/ResultReturner.java:112]: what's with the changes in this file? Doesn't seem related, this method for example isn't called in this diff.

**Comment by leaumar** [comment]
Reviewed with a relatively critical mindset, but I'm not a postgres guy so I'm unaware of some conventions that may or may not exist among postgresers.

File: postgres/src/main/java/org/jdbi/v3/postgres/ClobReaderColumnMapperFactory.java (hunk 2 above)
Comment (stevenschlansker) [reply postgres/src/main/java/org/jdbi/v3/postgres/ClobReaderColumnMapperFactory.java:21]: Yeah, supporting other encodings is left as a future enhancement.

File: postgres/src/test/java/org/jdbi/v3/postgres/TestLobStream.java (hunk 6 above)
Comment (leaumar) [reply postgres/src/test/java/org/jdbi/v3/postgres/TestLobStream.java:77]: I mean, the test limits the bounds to [0, 255], but everything about the impl says ints, so...

File: sqlobject/src/main/java/org/jdbi/v3/sqlobject/statement/internal/ResultReturner.java (hunk 9 above)
Comment (stevenschlansker) [reply sqlobject/src/main/java/org/jdbi/v3/sqlobject/statement/internal/ResultReturner.java:112]: Yeah, this was part of a refactoring that turned out not to be necessary.

File: postgres/src/test/java/org/jdbi/v3/postgres/TestLobStream.java (hunk 7 above)
Comment (stevenschlansker) [reply postgres/src/test/java/org/jdbi/v3/postgres/TestLobStream.java:76]: It's buffered in the driver so the test only takes a few seconds to run.

File: postgres/src/test/java/org/jdbi/v3/postgres/TestLobStream.java (hunk 6 above)
Comment (stevenschlansker) [reply postgres/src/test/java/org/jdbi/v3/postgres/TestLobStream.java:77]: Yeah, `InputStream` works on bytes but the type is `int` since its *unsigned* bytes.

File: postgres/src/test/java/org/jdbi/v3/postgres/TestLobStream.java (hunk 5 above)
Comment (stevenschlansker) [reply postgres/src/test/java/org/jdbi/v3/postgres/TestLobStream.java:35]: It's just a big amount of data that would cause problems to use inline.  Nothing more.

File: postgres/src/main/java/org/jdbi/v3/postgres/BlobInputStreamArgumentFactory.java (hunk 1 above)
Comment (stevenschlansker) [reply postgres/src/main/java/org/jdbi/v3/postgres/BlobInputStreamArgumentFactory.java:13]: It only does that for now.  We can always change the name later if the class ends up doing more, it's not API.

File: postgres/src/test/java/org/jdbi/v3/postgres/TestLobStream.java (hunk 6 above)
Comment (leaumar) [reply postgres/src/test/java/org/jdbi/v3/postgres/TestLobStream.java:77]: Sure, but from a functional perspective this impl reads like two different books. Might as well use the `byte` type and name everything as such

File: postgres/src/main/java/org/jdbi/v3/postgres/PgLobImpl.java (hunk 3 above)
Comment (stevenschlansker) [reply postgres/src/main/java/org/jdbi/v3/postgres/PgLobImpl.java:13]: No, it's just picked as a "big enough" value.  I aligned it to the Postgres driver which uses 4K.
It's hard coded in the driver so we don't really have the ability to tune it properly, so we shouldn't expose configuring as API at this time.

**Comment by stevenschlansker** [comment]
Thanks for the review.  I think I covered most of the important bits.

File: postgres/src/test/java/org/jdbi/v3/postgres/TestLobStream.java (hunk 7 above)
Comment (leaumar) [reply postgres/src/test/java/org/jdbi/v3/postgres/TestLobStream.java:76]: I meant the assertions themselves, all the assertj internals churning. Wouldn't it make sense to shave off some seconds by asserting larger chunks?

File: postgres/src/test/java/org/jdbi/v3/postgres/TestLobStream.java (hunk 5 above)
Comment (leaumar) [reply postgres/src/test/java/org/jdbi/v3/postgres/TestLobStream.java:35]: Can we give it a serious name expressing that then instead of a cheeky one?

File: postgres/src/main/java/org/jdbi/v3/postgres/BlobInputStreamArgumentFactory.java (hunk 1 above)
Comment (leaumar) [reply postgres/src/main/java/org/jdbi/v3/postgres/BlobInputStreamArgumentFactory.java:13]: I meant if InputStreams couldn't serve purposes not implemented by this class, like different kinds of lobs or something.

File: postgres/src/test/java/org/jdbi/v3/postgres/TestLobStream.java (hunk 7 above)
Comment (stevenschlansker) [reply postgres/src/test/java/org/jdbi/v3/postgres/TestLobStream.java:76]: It's essentially nothing, the expensive resource is server round-trips and those are fixed by the driver.

File: postgres/src/main/java/org/jdbi/v3/postgres/BlobInputStreamArgumentFactory.java (hunk 1 above)
Comment (stevenschlansker) [reply postgres/src/main/java/org/jdbi/v3/postgres/BlobInputStreamArgumentFactory.java:13]: If / when that happens we might have to differentiate by a qualifier annotation?

File: postgres/src/test/java/org/jdbi/v3/postgres/TestLobStream.java (hunk 6 above)
Comment (stevenschlansker) [reply postgres/src/test/java/org/jdbi/v3/postgres/TestLobStream.java:77]: You can't use `byte` because it's a signed byte -- `-128..127` -- `InputStream` requires you to work in `0..255` which cannot fit in a Java `byte`.  This isn't a Jdbi-ism, this is a `java.io`-ism that we can't change.

Comment (stevenschlansker) [reply postgres/src/test/java/org/jdbi/v3/postgres/TestLobStream.java:77]: Also, the return value has a special flag value `-1` for EOF which could not also fit in a byte

File: postgres/src/test/java/org/jdbi/v3/postgres/TestLobStream.java (hunk 5 above)
Comment (stevenschlansker) [reply postgres/src/test/java/org/jdbi/v3/postgres/TestLobStream.java:35]: Coming up with fun names is part of the otherwise not-fun of writing tests... I'll change this I guess but it really is just a "big" number for testing purposes.  Consider focusing your reviewing time on the meat of the issue at hand.  If you're not comfortable with the api, maybe checkout the upstream docs and learn a bit about it!  I didn't know anything about the Postgres LargeObject stuff before starting this PR.
